### DISCUSSION TITLE
Azure OAuth Authentication Issues

### DISCUSSION THREAD

**Discussion by ghost** [ngoduykhanh/PowerDNS-Admin#1512] [Q&A] [answered]
Hi,

Anybody has Azure authentication working? I can't seem to get it working. Have an application in Azure, exposed API. Set the scope. I do get an azure login screen but fails with useless error code in the addressbar and webinterface itself sais 500error

Flask gives;
Apr 07 17:11:01 opvitdcpwebadmin001 gunicorn[314628]:     return render_template('errors/500.html', code=500, message=e), 500
Apr 07 17:11:01 opvitdcpwebadmin001 gunicorn[314628]:   File "/var/www/html/pdns/flask/lib/python3.9/site-packages/flask/templating.py", line 153, in render_template
Apr 07 17:11:01 opvitdcpwebadmin001 gunicorn[314628]:     ctx.app.update_template_context(context)
Apr 07 17:11:01 opvitdcpwebadmin001 gunicorn[314628]:   File "/var/www/html/pdns/flask/lib/python3.9/site-packages/flask/app.py", line 750, in update_template_context
Apr 07 17:11:01 opvitdcpwebadmin001 gunicorn[314628]:     context.update(func())
Apr 07 17:11:01 opvitdcpwebadmin001 gunicorn[314628]:   File "/var/www/html/pdns/flask/lib/python3.9/site-packages/flask_login/utils.py", line 407, in _user_context_processor
Apr 07 17:11:01 opvitdcpwebadmin001 gunicorn[314628]:     return dict(current_user=_get_user())
Apr 07 17:11:01 opvitdcpwebadmin001 gunicorn[314628]:   File "/var/www/html/pdns/flask/lib/python3.9/site-packages/flask_login/utils.py", line 372, in _get_user
Apr 07 17:11:01 opvitdcpwebadmin001 gunicorn[314628]:     current_app.login_manager._load_user()
Apr 07 17:11:01 opvitdcpwebadmin001 gunicorn[314628]:   File "/var/www/html/pdns/flask/lib/python3.9/site-packages/flask_login/login_manager.py", line 362, in _load_user
Apr 07 17:11:01 opvitdcpwebadmin001 gunicorn[314628]:     user_id = session.get("_user_id")
Apr 07 17:11:01 opvitdcpwebadmin001 gunicorn[314628]: AttributeError: 'NoneType' object has no attribute 'get'

**Comment by AzorianMatt** [comment, marked as answer]
@Hentheroot I actually just got my local dev setup working with Azure pretty easily in the last week. I can't speak for production environments, but I setup a developer account with an Azure sandbox environment loaded with test data. Once I applied the appropriate settings in the authentication settings area of the app, it just worked. Try upgrading to the latest `0.4.1` release and implementing based on the workflow in the view now.

**Comment by ghost** [comment]
oh nice, trying it now. Problem now is that the authorized url is not correct.
As you can see it's somewhere defined as 'http://'  , so not with httpS://.  But where is this defined? I cannot find it.
powerdnsadmin works fully on SSL certificates. So I don't understand where it gets this 'http://' from.

![image](https://user-images.githubusercontent.com/79045580/231446160-a0efedfe-4e7a-43a8-bcfd-e5b59909158d.png)

**Reply by AzorianMatt** [reply]
@Hentheroot if your installation is behind a reverse proxy that provides SSL, try setting the new environment setting `SERVER_EXTERNAL_SSL` to a value of "1". My changes included support to utilize oauth locally without SSL so this setting must be enabled to override the detection of the local environments HTTP schema.

**Reply by AzorianMatt** [reply]
@Hentheroot I'm updating the `0.4.1` release to include a better default value for the `SERVER_EXTERNAL_SSL` setting so that this confusion doesn't come up for the standard PDA user. This has been updated in the `master` branch and the Docker image is building now. If you pull a fresh `0.4.1` image in a bit, you should have this included automatically which should correct your issue.

**Comment by ghost** [comment]
Ah nice, it works!!!  It indeed was behind a proxy. Thank you so much for this Azure fix. Was starting to pull my hairs out.

**Reply by AzorianMatt** [reply]
@Hentheroot excellent! You're welcome and thank you for being a PDA user!

**Comment by reckart-everstream** [comment]
I'm experiencing something similar to this and I'm running v0.4.2, got any suggestions @AzorianMatt 
File "/opt/web/powerdns-admin/venv/lib/python3.10/site-packages/flask/app.py", line 2529, in wsgi_app
Apr 24 19:26:48 ns1 gunicorn[52881]:     response = self.full_dispatch_request()
Apr 24 19:26:48 ns1 gunicorn[52881]:   File "/opt/web/powerdns-admin/venv/lib/python3.10/site-packages/flask/app.py", line 1825, in full_dispatch_request
Apr 24 19:26:48 ns1 gunicorn[52881]:     rv = self.handle_user_exception(e)
Apr 24 19:26:48 ns1 gunicorn[52881]:   File "/opt/web/powerdns-admin/venv/lib/python3.10/site-packages/flask/app.py", line 1823, in full_dispatch_request
Apr 24 19:26:48 ns1 gunicorn[52881]:     rv = self.dispatch_request()
Apr 24 19:26:48 ns1 gunicorn[52881]:   File "/opt/web/powerdns-admin/venv/lib/python3.10/site-packages/flask/app.py", line 1799, in dispatch_request
Apr 24 19:26:48 ns1 gunicorn[52881]:     return self.ensure_sync(self.view_functions[rule.endpoint])(**view_args)
Apr 24 19:26:48 ns1 gunicorn[52881]:   File "/opt/web/powerdns-admin/powerdnsadmin/routes/index.py", line 237, in login
Apr 24 19:26:48 ns1 gunicorn[52881]:     azure_username = user_data["userPrincipalName"]
Apr 24 19:26:48 ns1 gunicorn[52881]: KeyError: 'userPrincipalName'
